### ISSUE TITLE
Error connecting to voice on Raspberry Pi

### ISSUE THREAD

**Issue by JakeStanger** [DV8FromTheWorld/JDA#47]
I haven't had time to test this on another Linux device yet, but I know my code works fine on Windows.

When my bot tries to connect to voice, although actually connecting to the channel succeeds, opening the audio connection fails.

The following error occurs:

```
java.lang.NullPointerException
        at net.dv8tion.jda.audio.AudioWebSocket.send(AudioWebSocket.java:110)
        at net.dv8tion.jda.audio.AudioWebSocket.onConnected(AudioWebSocket.java:124)
        at com.neovisionaries.ws.client.ListenerManager.callOnConnected(ListenerManager.java:189)
        at com.neovisionaries.ws.client.WebSocket.callOnConnectedIfNotYet(WebSocket.java:3207)
        at com.neovisionaries.ws.client.WebSocket.onReadingThreadStarted(WebSocket.java:3152)
        at com.neovisionaries.ws.client.ReadingThread.main(ReadingThread.java:80)
        at com.neovisionaries.ws.client.ReadingThread.run(ReadingThread.java:61)
Exception in thread "Thread-5" Exception in thread "Thread-6" java.lang.NullPointerException
        at net.dv8tion.jda.audio.AudioConnection$3.run(AudioConnection.java:224)
java.lang.NullPointerException
        at net.dv8tion.jda.audio.AudioConnection$2.run(AudioConnection.java:151)
```

Other than that, I'd like to say that this API is brilliant and really easy to use, so thanks for all the hard work.

**Comment by DV8FromTheWorld** [comment]
If it had something to do with the native libraries it would have exploded long before this.
Anyways, what version are you running?
Also give me some specs about the Pi and OS you are running.

Lastly, once again please confirm that what you have works on windows and also test on another linux device if possible.

**Comment by JakeStanger** [comment]
Sorry, should have included specs.
Running v2.0.0 on Rpi model A (arm v6 cpu and 256MB RAM) running Raspbian Wheezy (from mid 2015, Jesse doesn't fit on my SD).

I cannot currently test it on Windows (I'm running 10) or another Linux version (I'll be able to test Ubuntu 15.10), but I did perform several tests on Windows before my Pi.

See [here](https://github.com/JakeStanger/Discord-Bot/blob/master/src/main/java/bot/Bot.java#L173) for my code if it's of any significance

**Comment by DV8FromTheWorld** [comment]
What specific build of v2.0 are you running?

**Comment by JakeStanger** [comment]
JDA-withDependencies-2.0.0_242.jar, which is the latest build.

I've got a feeling this might be an issue with the JDA connecting twice in my code, but I'm not in a position where I can test that right now.

**Comment by JakeStanger** [comment]
Ok so I've done several tests, and here's where I'm at.

-Windows 10 it works absolutely flawlessly.
-Raspbian I get a NullPointerException opening/closing the voice connection.
-Ubuntu the voice connection doesn't show any errors on connection/disconnection, but I haven't been able to test sounds due to a few other errors on my part. This would suggest it works though...


Here's the (similar) exception for closing a voice connection (possibly just because it was never opened in the first place):
`[16:21:41] [Fatal] [JDA]: One of the EventListeners had an uncaught exception
[16:21:41] [Fatal] [JDA]: Encountered an exception:
java.lang.NullPointerException
        at net.dv8tion.jda.audio.AudioConnection.close(AudioConnection.java:136)
        at net.dv8tion.jda.managers.impl.AudioManagerImpl.closeAudioConnection(AudioManagerImpl.java:123)
        at bot.Bot.leaveChannel(Bot.java:190)
        at bot.Bot.onGuildMessageReceived(Bot.java:112)
        at net.dv8tion.jda.hooks.ListenerAdapter.onEvent(ListenerAdapter.java:158)
        at net.dv8tion.jda.hooks.InterfacedEventManager.handle(InterfacedEventManager.java:57)
        at net.dv8tion.jda.handle.MessageReceivedHandler.handleInternally(MessageReceivedHandler.java:52)
        at net.dv8tion.jda.handle.SocketHandler.handle(SocketHandler.java:36)
        at net.dv8tion.jda.requests.WebSocketClient.handleEvent(WebSocketClient.java:545)
        at net.dv8tion.jda.requests.WebSocketClient.onTextMessage(WebSocketClient.java:296)
        at com.neovisionaries.ws.client.ListenerManager.callOnTextMessage(ListenerManager.java:352)
        at com.neovisionaries.ws.client.ReadingThread.callOnTextMessage(ReadingThread.java:233)
        at com.neovisionaries.ws.client.ReadingThread.callOnTextMessage(ReadingThread.java:211)
        at com.neovisionaries.ws.client.ReadingThread.handleTextFrame(ReadingThread.java:910)
        at com.neovisionaries.ws.client.ReadingThread.handleFrame(ReadingThread.java:693)
        at com.neovisionaries.ws.client.ReadingThread.main(ReadingThread.java:102)
        at com.neovisionaries.ws.client.ReadingThread.run(ReadingThread.java:61)
Exception in thread "Thread-5" Exception in thread "Thread-6" java.lang.NullPointerException
        at net.dv8tion.jda.audio.AudioConnection$3.run(AudioConnection.java:224)
java.lang.NullPointerException
        at net.dv8tion.jda.audio.AudioConnection$2.run(AudioConnection.java:151)`

**Comment by FrozenDroid** [comment]
I'm getting this exact same error, on a dedicated server running Debian 8.4. Sometimes it works, sometimes it doesn't.

**Comment by DV8FromTheWorld** [comment]
This is really interesting. 
I haven't looked into it yet but I'm wondering if there is a race condition...

**Comment by JakeStanger** [comment]
It could well be; whereas JDA connects in about a second on my PC, on my Pi it takes closer to 10.

**Comment by DV8FromTheWorld** [comment]
Can you do me a favor and test this with `2.0.0_247`?

**Comment by JakeStanger** [comment]
It's working fine on 247, so that's great. Out of interest, what was the problem?

**Comment by DV8FromTheWorld** [comment]
https://github.com/DV8FromTheWorld/JDA/commit/329b3e231db43a14057d62e31a77240f75ea94c4
A race condition like I thought. Basically, the `onConnect` method was being called during the `connect()` method's execution. Connect runs in the thread it is called from, but the `onConnect` method is handled by the websocket thread, which is a different thread. This means that the `onConnect` can be executed before `connect()` returns. We were using the return from `connect()` to populate our `socket` variable which is used in the `send(String)` method. So, since `connect()` hadn't returned, using the `socket` variable, which was still null, was throwing an NPE.

I changed it so that we set the `socket` variable before attempting to connect. We had  the same problem with the Main WebSocket, this specific race condition. However, it was a LOOONG time ago, so I guess I forgot to implement the fix in the AudioWebSocket when I created it.
